### PR TITLE
Improve `<SvgIcon>` to make it output `svg` node and optimize performance

### PR DESCRIPTION
**Pull request by wxiaoguang**
Before, the Vue `<SvgIcon>` always outputs DOM nodes like: 

```html
<span class="outer-class">
    <svg class="class-name-defined" ...></svg>
</span>
```

The `span` is redundant  and I guess such layout and the inconsistent `class/class-name` attributes would cause bugs sooner or later.

This PR makes the `<SvgIcon>` clear, and it's faster than before, because it doesn't need to parse the whole SVG string.

Before: 

<details>

image

</details>

After:

image

### CODE REVIEW

File: web_src/js/svg.js
Comment (silverwind) [inline web_src/js/svg.js:151]: What's the `^` for?

**Comment by silverwind** [comment]
This does seem like a lot of extra code and it redoes most of what `svg()` already does. I wonder if it wouldn't be better to have a shared parser function that returns `{attributes, innerHTML}` which could be used in both.

**Comment by wxiaoguang** [comment]
> This does seem like a lot of extra code and it redoes most of what svg already does.

Nope, not extra.

>  I wonder if it wouldn't be better to have a shared parser function that returns {attributes, innerHTML} which could be used in both. 

They do not the same thing. The key point is, in your old `svg` function, the string is returned for default arguments.

> It would lose the fast-path of svg with size 16 and no classes, thought, so that is a downside and probably not worth because of that.

No, nothing is lost.

Before, browser parse both the svgOuterHtml and svgInnerHtml

Now, we parse the svgOuterHtml, and leave the svgInnerHtml to browser.

Nothing is parsed twice.

File: web_src/js/svg.js
Comment (wxiaoguang) [reply web_src/js/svg.js:151]: There is a comment above.

**Comment by wxiaoguang** [comment]
Actually, the old SvgIcon is slower, because it does parse twice for non-default arguments.

* It parses the whole SVG string to Node (svgOuterHtml, svgInnerHtml)
* It converts the Node to string (then svgOuterHtml and svgInnerHtml are parsed again by the browser)

After this PR:

* This PR first parses svgOuterHtml, it doesn't parse svgInnerHtml
* Then it creates a VNode by svgOuterHtml parsed node, which is faster
* Leave the svgInnerHtml  to browser, which is faster.
* Nothing is parsed twice.

File: web_src/js/svg.js
Comment (silverwind) [reply web_src/js/svg.js:151]: Hmm I still don't get it, that link does not explain it. Why can we have `id` or `class` but `^height` needs the `^`? These vue docs seem incomplete, there is also apparently a `.` prefix.

Comment (wxiaoguang) [reply web_src/js/svg.js:151]: `.` is for `prop`, eg: `el.disabled` / `el.textContent` -> `patchDOMProp`

`^` is for attribute, eg: `<tag data-xxx=xxx>`, `<tag width=xxx>` -> `patchAttr`

The code is here

https://github.com/vuejs/core/blob/c6eb3ccccee8e43e7aafcbc3c9ededecc565fdf0/packages/runtime-dom/src/patchProp.ts

It's a little difficult to understand.

Comment (wxiaoguang) [reply web_src/js/svg.js:151]: These prefixes are used to avoid conflicting. Without these prefixes, Vue would "guess" what it is.

With these prefixes, it's explicitly defined that these are attributes.

----

That's what the document means: 

> // props modifiers such as `.prop` and `.attr` can be added
> // with `.` and `^` prefixes respectively

Comment (silverwind) [reply web_src/js/svg.js:151]: Understood. Vue is even worse then React when it comes to mangling DOM attribute names, but ultimately it's a failure of the DOM standard to even have a difference between a prop and an attr.

File: web_src/js/svg.js
Comment (HesterG) [inline web_src/js/svg.js:123]: Looks like span is used in some css selectors
截屏2023-03-21 14 56 07

So if span is removed, on some pages like action can't match the rule (in the case below, no green. For now I only find [this](https://github.com/go-gitea/gitea/blob/a797b8458e90caa00427d1270b82031c5e64ec09/web_src/css/base.css#L2423-L2434) though)
截屏2023-03-21 14 58 16

Comment (wxiaoguang) [reply web_src/js/svg.js:123]: IMO that's a kind of side-effect, not that good. I think these usages should be fixed, instead of making it more unclear.

And these styles are old, I will try to make some fine tunes.

image

Comment (wxiaoguang) [reply web_src/js/svg.js:123]: OK, I am 99% sure there were mis-uses of these styles.

That's just what I said:

> The span is redundant and I guess such layout and the inconsistent class/class-name attributes would cause bugs sooner or later.

The SVG color should be provided by ".text.color", but not by "span.color .svg", see  https://github.com/go-gitea/gitea/pull/23570/commits/367ffca1961a447d692406617ddcc09a69f0eb12


So I'd like to remove these "span.color .svg" styles, and fix incorrect UI elements from now on.

Comment (wxiaoguang) [reply web_src/js/svg.js:123]: There were some other incorrect usages, @silverwind has fixed them in #23565, so I won't touch the `notification_div.tmpl` in this PR

**Review by HesterG — APPROVED** [review]

File: web_src/js/svg.js
Comment (silverwind) [reply web_src/js/svg.js:123]: Yes, try to get rid of remaining use of these selectors and the nuke them 👍

File: web_src/js/svg.js
Comment (silverwind) [inline web_src/js/svg.js]: Can we remove this string matching? We are already parsing the document so assertions can be done on the parsed result.

Comment (wxiaoguang) [reply web_src/js/svg.js]: Where? There is no parsed result.

Comment (wxiaoguang) [reply web_src/js/svg.js]: To be clear, here, we only parse the OuterHtml, to make it fast.

We never parse the InnerHtml, we just leave it to browser.

Comment (silverwind) [reply web_src/js/svg.js]: I mean remove lines 130,131 and try/catch around the `parseFromString` instead. Faster and catches more kind of errors.

Comment (wxiaoguang) [reply web_src/js/svg.js]: I do not think it's worth to parse the whole SVG string. We do not need to pay the cost for its inner.

Comment (silverwind) [reply web_src/js/svg.js]: Actually you don't even need to split inner and outer via string, just query into the parsed document.

Comment (wxiaoguang) [reply web_src/js/svg.js]: Where is the `parsed document` before you parse? When you parse, you spend time on its inner, but we never use its inner.

File: web_src/js/svg.js
Comment (silverwind) [inline web_src/js/svg.js]:
```suggestion
  if (size !== 16) {
    svgNode.setAttribute('width', String(size));
    svgNode.setAttribute('height', String(size));
  } 
```

Tiny optimization I just noticed.

File: web_src/js/svg.js
Comment (wxiaoguang) [reply web_src/js/svg.js]: OK, to be clear, let's say 2 approaches.

1. My approach: we only parse the Outer, never parse the inner. I think it's faster because we do not need to spend time on inner.
2. Your suggestion, parse the whole SVG string, including its outer and inner (slow path). Then we need to call "stringify(inner)" to get the HTML string for VNode.innerHTML , that's another slow path.

Do you really prefer your suggestion?

Comment (silverwind) [reply web_src/js/svg.js]: I'm just not sure how robust the string parsing is but given that `<` and `>` need to be escaped in XML, I think it's safe. Would suggest adding some tests for this function in [`svg.test.js`](https://github.com/go-gitea/gitea/blob/main/web_src/js/svg.test.js).

Comment (wxiaoguang) [reply web_src/js/svg.js]: All SVG are generated by your "generate-svg.js" , they are strictly clean , so that's absolute safe.

About "tests", I have no idea how to write a test for it ..... a normal SVG like `<svg>...</svg>`?

Comment (silverwind) [reply web_src/js/svg.js]: Import `SvgIcon` into the test file and probably also Vue. Then make Vue render to a temporary element or string and assert the output.

Comment (silverwind) [reply web_src/js/svg.js]: https://vuejs.org/api/ssr.html#rendertostring probably easiest with that API, thought not totally representative as we don't do SSR.

Comment (wxiaoguang) [reply web_src/js/svg.js]: Does https://github.com/go-gitea/gitea/pull/23570/commits/d30cc25d049ab4c113f44cf276e141d7197dbc8e look good to you ?

Comment (silverwind) [reply web_src/js/svg.js]: Would prefer to test the whole thing. Not an expert on Vue, but something like this maybe:

```js
import {SvgIcon} from './svg.js';
import {createApp} from 'vue';

test('SvgIcon', () => {
  const root = document.createElement('div');
  createApp({data: () => SvgIcon('octicon-link', 24, 'class-name')}).mount(root);
  expect(root.firstChild.nodeName).toEqual('SVG');
  expect(root.firstChild.height).toEqual(24);
  expect(root.firstChild.width).toEqual(24);
  expect(root.firstChild.className).toEqual('svg octicon-link');
});
```

Edit: Added `.mount()`

Comment (wxiaoguang) [reply web_src/js/svg.js]:
```
TypeError: Cannot read properties of null (reading 'nodeName')
```

`console.log(root.querySelectorAll('*').length);` says:  `0`

Comment (silverwind) [reply web_src/js/svg.js]: Maybe we need https://test-utils.vuejs.org/guide/#what-is-vue-test-utils

Comment (wxiaoguang) [reply web_src/js/svg.js]: > Edit: Added `.mount()`

```
SvgIcon is not a function
TypeError: __vite_ssr_import_1__.SvgIcon is not a function
```

You have my fork's write access, feel free to edit this PR directly.

Comment (silverwind) [reply web_src/js/svg.js]: Will check later, I feel like this should be possible without additional dependencies like this `@vue/test-utils` which I'd like to avoid.

Comment (silverwind) [reply web_src/js/svg.js]: Just for info, the reference example of vitest for vue uses test-utils, but we'll see if those are really necessary:

https://github.com/vitest-dev/vitest/blob/main/examples/vue/test/basic.test.ts

Comment (wxiaoguang) [reply web_src/js/svg.js]: Added 0b0082652d970e065782f87b02c04c01cedbbd10

```js
test('SvgIcon', () => {
  const root = document.createElement('div');
  createApp({render: () => h(SvgIcon, {name: 'octicon-link', size: 24})}).mount(root);
  expect(root.firstChild.nodeName).toEqual('svg');
  expect(root.firstChild.getAttribute('width')).toEqual('24');
  expect(root.firstChild.classList.contains('octicon-link')).toBeTruthy();
}
```

Comment (silverwind) [reply web_src/js/svg.js]: Nice, that should be it then.

**Review by silverwind — APPROVED** [review]

**Pull request merged**
